File: framework-internal/src/test/java/org/identityconnectors/framework/impl/api/local/operations/UpdateImplTests.java
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 * 
 * The contents of this file are subject to the terms of the Common Development 
 * and Distribution License("CDDL") (the "License").  You may not use this file 
 * except in compliance with the License.
 * 
 * You can obtain a copy of the License at 
 * http://IdentityConnectors.dev.java.net/legal/license.txt
 * See the License for the specific language governing permissions and limitations 
 * under the License. 
 * 
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at identityconnectors/legal/license.txt.
 * If applicable, add the following below this CDDL Header, with the fields 
 * enclosed by brackets [] replaced by your own identifying information: 
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
package org.identityconnectors.framework.impl.api.local.operations;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.identityconnectors.common.CollectionUtil;
import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.AttributeBuilder;
import org.identityconnectors.framework.common.objects.Name;
import org.identityconnectors.framework.common.objects.ObjectClass;
import org.identityconnectors.framework.common.objects.Uid;
import org.identityconnectors.framework.impl.api.local.operations.UpdateImpl;
import org.junit.Assert;
import org.junit.Test;


/**
 * Testing for the merging of the various attribute during update.
 */
public class UpdateImplTests {

    @Test(expected=NullPointerException.class)
    public void validateUidArg() {
        UpdateImpl.validateInput(ObjectClass.ACCOUNT, null, new HashSet<Attribute>(),true);
    }

    @Test(expected=NullPointerException.class)
    public void validateObjectClassArg() {
        UpdateImpl.validateInput(null, new Uid("foo"), new HashSet<Attribute>(),true);
    }
    
    @Test(expected=NullPointerException.class)
    public void validateAttrsArg() {
        UpdateImpl.validateInput(ObjectClass.ACCOUNT, new Uid("foo"),null,true);
    }
    
    @Test(expected=IllegalArgumentException.class)
    public void validateUidAttribute() {
        Set<Attribute> attrs=new HashSet<Attribute>();
        attrs.add(new Uid("foo"));
        UpdateImpl.validateInput(ObjectClass.ACCOUNT, new Uid("foo"),attrs,true);
    }
    
    @Test(expected=IllegalArgumentException.class)
    public void validateAddWithNullAttribute() {
        Set<Attribute> attrs = new HashSet<Attribute>();
        attrs.add(AttributeBuilder.build("something"));
        UpdateImpl.validateInput(ObjectClass.ACCOUNT, new Uid("foo"), attrs, true);        
    }
    
    @Test(expected=IllegalArgumentException.class)
    public void validateAttemptToAddName() {
        Set<Attribute> attrs = new HashSet<Attribute>();
        attrs.add(new Name("fadf"));
        UpdateImpl.validateInput(ObjectClass.ACCOUNT, new Uid("foo"),attrs,true);                
    }

    @Test
    public void validateAttemptToAddDeleteOperationalAttribute() {
        // list of all the operational attributes..
        List<Attribute> list = new ArrayList<Attribute>();
        list.add(AttributeBuilder.buildEnabled(false));
        list.add(AttributeBuilder.buildLockOut(true));
        list.add(AttributeBuilder.buildCurrentPassword("fadsf".toCharArray()));
        list.add(AttributeBuilder.buildPasswordExpirationDate(new Date()));
        list.add(AttributeBuilder.buildPassword("fadsf".toCharArray()));
        for (Attribute attr : list) {
            Set<Attribute> attrs = new HashSet<Attribute>();
            attrs.add(attr);
            try {
                UpdateImpl.validateInput(ObjectClass.ACCOUNT, new Uid("1"), attrs,true);
                Assert.fail("Failed: " + attr.getName());
            } catch (IllegalArgumentException e) {
                // this is a good thing..
            }
        }
    }
    
    @Test
    public void mergeAddAttribute() {
        UpdateImpl up = new UpdateImpl(null, null);
        Set<Attribute> actual;
        Set<Attribute> base = CollectionUtil.<Attribute>newSet();
        Set<Attribute> expected = CollectionUtil.<Attribute>newSet();
        Set<Attribute> changeset = CollectionUtil.<Attribute>newSet();
        // attempt to add a value to an attribute..
        Attribute cattr = AttributeBuilder.build("abc", 2);
        changeset.add(cattr);
        expected.add(AttributeBuilder.build("abc", 2));        
        actual = up.merge(changeset, base, true);
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void mergeAddToExistingAttribute() {
        UpdateImpl up = new UpdateImpl(null, null);
        Set<Attribute> actual;
        Set<Attribute> base = CollectionUtil.<Attribute>newSet();
        Set<Attribute> expected = CollectionUtil.<Attribute>newSet();
        Set<Attribute> changeset = CollectionUtil.<Attribute>newSet();
        // attempt to add a value to an attribute..
        Attribute battr = AttributeBuilder.build("abc", 1);
        Attribute cattr = AttributeBuilder.build("abc", 2);
        base.add(battr);
        changeset.add(cattr);
        expected.add(AttributeBuilder.build("abc", 1, 2));        
        actual = up.merge(changeset, base,true);
        Assert.assertEquals(expected, actual);
    }
    
    @Test
    public void mergeDeleteNonExistentAttribute() {
        UpdateImpl up = new UpdateImpl(null, null);
        Set<Attribute> actual;
        Set<Attribute> base = CollectionUtil.<Attribute>newSet();
        Set<Attribute> expected = CollectionUtil.<Attribute>newSet();
        Set<Attribute> changeset = CollectionUtil.<Attribute>newSet();
        // attempt to add a value to an attribute..
        Attribute cattr = AttributeBuilder.build("abc", 2);
        changeset.add(cattr);
        actual = up.merge(changeset, base, false);
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void mergeDeleteToExistingAttribute() {
        UpdateImpl up = new UpdateImpl(null, null);
        Set<Attribute> actual;
        Set<Attribute> base = CollectionUtil.<Attribute>newSet();
        Set<Attribute> expected = CollectionUtil.<Attribute>newSet();
        Set<Attribute> changeset = CollectionUtil.<Attribute>newSet();
        // attempt to add a value to an attribute..
        Attribute battr = AttributeBuilder.build("abc", 1, 2);
        Attribute cattr = AttributeBuilder.build("abc", 2);
        base.add(battr);
        changeset.add(cattr);
        expected.add(AttributeBuilder.build("abc", 1));
        actual = up.merge(changeset, base, false);
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void mergeDeleteToExistingAttributeCompletely() {
        UpdateImpl up = new UpdateImpl(null, null);
        Set<Attribute> actual;
        Set<Attribute> base = CollectionUtil.<Attribute>newSet();
        Set<Attribute> expected = CollectionUtil.<Attribute>newSet();
        Set<Attribute> changeset = CollectionUtil.<Attribute>newSet();
        // attempt to add a value to an attribute..
        Attribute battr = AttributeBuilder.build("abc", 1, 2);
        Attribute cattr = AttributeBuilder.build("abc", 1, 2);
        base.add(battr);
        changeset.add(cattr);
        expected.add(AttributeBuilder.build("abc"));
        actual = up.merge(changeset, base, false);
        Assert.assertEquals(expected, actual);
    }

    
}

File: dotnet/FrameworkTests/UpdateImplTests.cs
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 * 
 * The contents of this file are subject to the terms of the Common Development 
 * and Distribution License("CDDL") (the "License").  You may not use this file 
 * except in compliance with the License.
 * 
 * You can obtain a copy of the License at 
 * http://opensource.org/licenses/cddl1.php
 * See the License for the specific language governing permissions and limitations 
 * under the License. 
 * 
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at http://opensource.org/licenses/cddl1.php.
 * If applicable, add the following below this CDDL Header, with the fields 
 * enclosed by brackets [] replaced by your own identifying information: 
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
using System;
using System.Security;
using NUnit.Framework;
using System.Collections.Generic;
using Org.IdentityConnectors.Common;
using Org.IdentityConnectors.Framework.Api;
using Org.IdentityConnectors.Framework.Api.Operations;
using Org.IdentityConnectors.Framework.Impl.Api.Local.Operations;
using Org.IdentityConnectors.Framework.Common.Objects;
namespace FrameworkTests
{
    /// <summary>
    /// Description of UpdateImplTests.
    /// </summary>
    [TestFixture]
    public class UpdateImplTests
    {
        [Test]
        [ExpectedException(typeof(ArgumentNullException))]
        public void ValidateUidArg()
        {
            UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, null, new HashSet<ConnectorAttribute>(), true);
        }
        [Test]
        [ExpectedException(typeof(ArgumentNullException))]
        public void ValidateObjectClassArg()
        {
            UpdateImpl.ValidateInput(null, new Uid("foo"), new HashSet<ConnectorAttribute>(), true);
        }

        [Test]
        [ExpectedException(typeof(ArgumentNullException))]
        public void ValidateAttrsArg()
        {
            UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, new Uid("foo"), null, true);
        }

        [Test]
        [ExpectedException(typeof(ArgumentException))]
        public void ValidateUidAttribute()
        {
            HashSet<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
            attrs.Add(new Uid("foo"));
            UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, new Uid("foo"), attrs, true);
        }

        [Test]
        [ExpectedException(typeof(ArgumentException))]
        public void ValidateAddWithNullAttribute()
        {
            ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
            attrs.Add(ConnectorAttributeBuilder.Build("something"));
            UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, new Uid("foo"), attrs, true);
        }

        [Test]
        [ExpectedException(typeof(ArgumentException))]
        public void ValidateAttemptToAddName()
        {
            ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
            attrs.Add(new Name("fadf"));
            UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, new Uid("foo"), attrs, true);
        }

        [Test]
        public void ValidateAttemptToAddDeleteOperationalAttribute()
        {
            // list of all the operational attributes..
            ICollection<ConnectorAttribute> list = new List<ConnectorAttribute>();
            list.Add(ConnectorAttributeBuilder.BuildEnabled(false));
            list.Add(ConnectorAttributeBuilder.BuildLockOut(true));
            list.Add(ConnectorAttributeBuilder.BuildCurrentPassword(newSecureString("fadsf")));
            list.Add(ConnectorAttributeBuilder.BuildPasswordExpirationDate(DateTime.Now));
            list.Add(ConnectorAttributeBuilder.BuildPassword(newSecureString("fadsf")));
            foreach (ConnectorAttribute attr in list)
            {
                ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
                attrs.Add(attr);
                try
                {
                    UpdateImpl.ValidateInput(ObjectClass.ACCOUNT, new Uid("1"), attrs, true);
                    Assert.Fail("Failed: " + attr.Name);
                }
                catch (ArgumentException)
                {
                    // this is a good thing..
                }
            }
        }

        private static SecureString newSecureString(string password)
        {
            SecureString rv = new SecureString();
            foreach (char c in password.ToCharArray())
            {
                rv.AppendChar(c);
            }
            return rv;
        }

        /// <summary>
        /// Validate two collections are equal.  (Not fast but effective)
        /// </summary>
        public static bool AreEqual(ICollection<ConnectorAttribute> arg1,
                                    ICollection<ConnectorAttribute> arg2)
        {
            if (arg1.Count != arg2.Count)
            {
                return false;
            }
            foreach (ConnectorAttribute attr in arg1)
            {
                if (!arg2.Contains(attr))
                {
                    return false;
                }
            }
            return true;
        }
        [Test]
        public void MergeAddAttribute()
        {
            UpdateImpl up = new UpdateImpl(null, null);
            ICollection<ConnectorAttribute> actual;
            ICollection<ConnectorAttribute> baseAttrs = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> expected = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> changeset = CollectionUtil.NewSet<ConnectorAttribute>();
            // attempt to add a value to an attribute..
            ConnectorAttribute cattr = ConnectorAttributeBuilder.Build("abc", 2);
            changeset.Add(cattr);
            expected.Add(ConnectorAttributeBuilder.Build("abc", 2));
            actual = up.Merge(changeset, baseAttrs, true);
            Assert.IsTrue(AreEqual(expected, actual));
        }

        [Test]
        public void MergeAddToExistingAttribute()
        {
            UpdateImpl up = new UpdateImpl(null, null);
            ICollection<ConnectorAttribute> actual;
            ICollection<ConnectorAttribute> baseAttrs = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> expected = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> changeset = CollectionUtil.NewSet<ConnectorAttribute>();
            // attempt to add a value to an attribute..
            ConnectorAttribute battr = ConnectorAttributeBuilder.Build("abc", 1);
            ConnectorAttribute cattr = ConnectorAttributeBuilder.Build("abc", 2);
            baseAttrs.Add(battr);
            changeset.Add(cattr);
            expected.Add(ConnectorAttributeBuilder.Build("abc", 1, 2));
            actual = up.Merge(changeset, baseAttrs, true);
            Assert.IsTrue(AreEqual(expected, actual));
        }

        [Test]
        public void MergeDeleteNonExistentAttribute()
        {
            UpdateImpl up = new UpdateImpl(null, null);
            ICollection<ConnectorAttribute> actual;
            ICollection<ConnectorAttribute> baseAttrs = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> expected = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> changeset = CollectionUtil.NewSet<ConnectorAttribute>();
            // attempt to add a value to an attribute..
            ConnectorAttribute cattr = ConnectorAttributeBuilder.Build("abc", 2);
            changeset.Add(cattr);
            actual = up.Merge(changeset, baseAttrs, false);
            Assert.IsTrue(AreEqual(expected, actual));
        }

        [Test]
        public void MergeDeleteToExistingAttribute()
        {
            UpdateImpl up = new UpdateImpl(null, null);
            ICollection<ConnectorAttribute> actual;
            ICollection<ConnectorAttribute> baseAttrs = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> expected = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> changeset = CollectionUtil.NewSet<ConnectorAttribute>();
            // attempt to add a value to an attribute..
            ConnectorAttribute battr = ConnectorAttributeBuilder.Build("abc", 1, 2);
            ConnectorAttribute cattr = ConnectorAttributeBuilder.Build("abc", 2);
            baseAttrs.Add(battr);
            changeset.Add(cattr);
            expected.Add(ConnectorAttributeBuilder.Build("abc", 1));
            actual = up.Merge(changeset, baseAttrs, false);
            Assert.IsTrue(AreEqual(expected, actual));
        }

        [Test]
        public void MergeDeleteToExistingAttributeCompletely()
        {
            UpdateImpl up = new UpdateImpl(null, null);
            ICollection<ConnectorAttribute> actual;
            ICollection<ConnectorAttribute> baseAttrs = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> expected = CollectionUtil.NewSet<ConnectorAttribute>();
            ICollection<ConnectorAttribute> changeset = CollectionUtil.NewSet<ConnectorAttribute>();
            // attempt to add a value to an attribute..
            ConnectorAttribute battr = ConnectorAttributeBuilder.Build("abc", 1, 2);
            ConnectorAttribute cattr = ConnectorAttributeBuilder.Build("abc", 1, 2);
            baseAttrs.Add(battr);
            changeset.Add(cattr);
            expected.Add(ConnectorAttributeBuilder.Build("abc"));
            actual = up.Merge(changeset, baseAttrs, false);
            Assert.IsTrue(AreEqual(expected, actual));
        }
    }
}
